libjpeg_addons/src/libjpg_addons.c

```diff
@@ -0,0 +1,294 @@
+{
+	jpgData *ret;
+	struct jpeg_decompress_struct cinfo;
+	struct my_error_mgr jerr;
+
+	cinfo.err = jpeg_std_error(&jerr.pub);
+	jerr.pub.error_exit = my_error_exit;
+	if (setjmp(jerr.setjmp_buffer)) {
+		jpeg_destroy_decompress(&cinfo);
+		printf("jpgFromRAW: error during processing file\n");
+		return NULL;
+	}
+	jpeg_create_decompress(&cinfo);
+	jpeg_mem_src(&cinfo, (void *)data, size);
+	jpeg_read_header(&cinfo, TRUE);
+
+	ret = _load_JPEG(&cinfo, mode);
+	
+	jpeg_destroy_decompress(&cinfo);
+
+	return ret;
+}
+
+jpgData *jpgFromFilename(const char *filename, int mode)
+{
+	void *data;
+	int size;
+	int fd;
+
+	fd = open(filename, O_RDONLY, 0);
+	if(fd == -1) {
+		printf("jpgOpen: error opening '%s'\n", filename);
+		return NULL;
+	}
+
+	size = lseek(fd, 0, SEEK_END);
+	lseek(fd, 0, SEEK_SET);
+
+	data = malloc(size);
+	if(data == NULL) {
+		printf("jpgOpen: NOT enought memory\n");
+		close(fd);
+		return NULL;
+	}
+	read(fd, data, size);
+	close(fd);
+
+	return jpgFromRAW(data, size, mode);
+}
+
+jpgData *jpgFromFILE(FILE *in_file, int mode)
+{
+	jpgData *jpg;
+	struct jpeg_decompress_struct cinfo;
+	struct my_error_mgr jerr;
+
+	cinfo.err = jpeg_std_error(&jerr.pub);
+	jerr.pub.error_exit = my_error_exit;
+	if (setjmp(jerr.setjmp_buffer)) {
+		jpeg_destroy_decompress(&cinfo);
+		printf("jpgOpenFILE: error during processing file\n");
+		return NULL;
+	}
+	jpeg_create_decompress(&cinfo);
+	jpeg_stdio_src(&cinfo, in_file);
+	jpeg_read_header(&cinfo, TRUE);
+
+	jpg = _load_JPEG(&cinfo, mode);
+	
+	jpeg_destroy_decompress(&cinfo);
+
+	return jpg;
+}
+
+void jpgFileFromJpgData(const char *filename, int quality, jpgData *jpg) {
+	struct jpeg_compress_struct cinfo;
+	struct my_error_mgr jerr;
+	FILE *outfile;
+
+	cinfo.err = jpeg_std_error(&jerr.pub);
+	jerr.pub.error_exit = my_error_exit;
+	if (setjmp(jerr.setjmp_buffer)) {
+		jpeg_destroy_compress(&cinfo);
+		printf("jpgOpenFILE: error during processing file\n");
+		return;
+	}
+	jpeg_create_compress(&cinfo);
+
+	if ((outfile = fopen(filename, "wb")) == NULL) {
+    	fprintf(stderr, "can't open %s\n", filename);
+    	return;
+  	}
+
+  	jpeg_stdio_dest(&cinfo, outfile);
+
+	cinfo.image_width = jpg->width;
+  	cinfo.image_height = jpg->height;
+  	cinfo.input_components = jpg->bpp/8;
+  	cinfo.in_color_space = JCS_RGB;
+  
+	jpeg_set_defaults(&cinfo);
+	jpeg_set_quality(&cinfo, 100, TRUE);
+
+
+	jpeg_start_compress(&cinfo, TRUE);
+
+	unsigned int row_stride = jpg->width * 3;
+	unsigned char *row_pointer = (unsigned char *)jpg->buffer;
+	while (cinfo.next_scanline < cinfo.image_height) {
+		jpeg_write_scanlines(&cinfo, (JSAMPARRAY)&row_pointer, 1);
+		row_pointer += row_stride;
+	}
+
+	jpeg_finish_compress(&cinfo);
+  	fclose(outfile);
+
+	jpeg_destroy_compress(&cinfo);
+}
+
+#define PS2SS_GSPSMCT32		0
+#define PS2SS_GSPSMCT24		1
+#define PS2SS_GSPSMCT16		2
+
+int jpgScreenshot(const char* filename,unsigned int vramAdress, unsigned int width, unsigned int height, unsigned int psm)
```

Comment:
With ps2 specific code in the library, should the library be named libjpeg_ps2_addons?

Reply:
Ok, I can update it

Reply:
Done